File: src/lib/sources.ts
// V2.2.6 — single source of truth for the 8 literature sources.
//
// Both Literature Search and Today's Feed (and the Settings "文献数据源管理"
// card) import this so the source list can never drift between modules again.
// Which sources are *enabled* is a global, backend-persisted toggle
// (get_enabled_sources / set_enabled_sources); an EMPTY enabled list means
// "all enabled" (matches the backend convention).

export interface SourceDef {
  value: string;
  /** Brand name — not localized (matches the source-badge labels). */
  label: string;
  /** i18n key under `sources.group_*` for the Settings grouping header. */
  groupKey: string;
}

export const ALL_SOURCES: SourceDef[] = [
  { value: "arxiv", label: "arXiv", groupKey: "sources.group_general" },
  { value: "semantic_scholar", label: "Semantic Scholar", groupKey: "sources.group_general" },
  { value: "pubmed", label: "PubMed", groupKey: "sources.group_general" },
  { value: "openalex", label: "OpenAlex", groupKey: "sources.group_general" },
  { value: "crossref", label: "Crossref", groupKey: "sources.group_general" },
  { value: "core", label: "CORE", groupKey: "sources.group_oa" },
  { value: "doaj", label: "DOAJ", groupKey: "sources.group_oa" },
  { value: "dblp", label: "DBLP", groupKey: "sources.group_cs" },
];

export const ALL_SOURCE_VALUES: string[] = ALL_SOURCES.map((s) => s.value);

/** Empty enabled list = all enabled (backend convention). */
export function isSourceEnabled(enabled: string[], value: string): boolean {
  return enabled.length === 0 || enabled.includes(value);
}

/** Resolve a stored enabled list to the concrete set of enabled source ids
 *  (expands the "empty = all" convention). */
export function resolveEnabledSources(enabled: string[]): string[] {
  return enabled.length === 0 ? [...ALL_SOURCE_VALUES] : enabled;
}

File: src-tauri/src/config/sources.rs
//! V2.2.6 — the single global "enabled literature sources" toggle.
//!
//! One place configures which of the 8 sources are queried; Literature Search
//! and Today's Feed both read this, so the two stay consistent (the Settings
//! "文献数据源管理" card is the only editor). Stored as JSON in the data
//! directory so it travels with the rest of the data on a data-dir move.
//!
//! Convention (matches the old `AppConfig.enabled_sources`): an EMPTY list
//! means "all sources enabled" (fresh install / never configured).

use std::path::PathBuf;

use serde::{Deserialize, Serialize};

const FILE: &str = "sources.json";

/// Canonical id list — the order is also the search/merge priority.
pub const ALL_SOURCES: [&str; 8] = [
    "arxiv",
    "semantic_scholar",
    "pubmed",
    "openalex",
    "crossref",
    "core",
    "dblp",
    "doaj",
];

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct SourcesConfig {
    /// Enabled source ids. EMPTY = all enabled.
    #[serde(default)]
    enabled: Vec<String>,
}

fn file_path(app: &tauri::AppHandle) -> PathBuf {
    crate::config::paths::data_root(app).join(FILE)
}

/// Enabled source ids; empty = all (missing file / parse error → empty).
pub fn load_enabled(app: &tauri::AppHandle) -> Vec<String> {
    match std::fs::read_to_string(file_path(app)) {
        Ok(t) => serde_json::from_str::<SourcesConfig>(&t)
            .map(|c| c.enabled)
            .unwrap_or_default(),
        Err(_) => Vec::new(),
    }
}

/// Persist the enabled source ids (unknown ids are dropped so the file can't
/// drift from `ALL_SOURCES`).
pub fn save_enabled(app: &tauri::AppHandle, enabled: Vec<String>) -> Result<(), String> {
    let enabled: Vec<String> = enabled
        .into_iter()
        .filter(|s| ALL_SOURCES.contains(&s.as_str()))
        .collect();
    let path = file_path(app);
    if let Some(dir) = path.parent() {
        std::fs::create_dir_all(dir).map_err(|e| format!("create data dir: {}", e))?;
    }
    let text = serde_json::to_string_pretty(&SourcesConfig { enabled })
        .map_err(|e| format!("serialize sources.json: {}", e))?;
    std::fs::write(&path, text).map_err(|e| format!("write sources.json: {}", e))?;
    Ok(())
}

/// Whether `name` should be queried, given the loaded enabled set
/// (empty set = everything on).
pub fn is_enabled(enabled: &[String], name: &str) -> bool {
    enabled.is_empty() || enabled.iter().any(|s| s == name)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_means_all_enabled() {
        let enabled: Vec<String> = Vec::new();
        for s in ALL_SOURCES {
            assert!(is_enabled(&enabled, s), "{s} should be on when empty");
        }
    }

    #[test]
    fn subset_only_enables_listed() {
        let enabled = vec!["arxiv".to_string(), "dblp".to_string()];
        assert!(is_enabled(&enabled, "arxiv"));
        assert!(is_enabled(&enabled, "dblp"));
        assert!(!is_enabled(&enabled, "pubmed"));
        assert!(!is_enabled(&enabled, "core"));
    }
}
